### PR TITLE
overlord/auth,store: add macaroon serialization/deserialization helpers

### PR DESCRIPTION
**Pull request by matiasb**
Adds macaroon package dependency, and introduces macaroon serialization/deserialization helpers (compatible with store macaroon serialization).
These helpers will be used to extract macaroon information (which will be needed when requesting the discharge macaroon in the updated flow), and also to be able to prepare the discharge macaroon for request (ie. bind the discharge macaroon to the root macaroon).

### CODE REVIEW

File: dependencies.tsv [hunk 1]
```diff
@@ -13,6 +13,7 @@ github.com/testing-cabal/subunit-go	git	00b258565a5cf3adaa24b68d31c9e6ec3d2cdbe7
 golang.org/x/crypto	git	60052bd85f2d91293457e8811b0cf26b773de469	2015-06-22T23:34:07Z
 golang.org/x/net	git	3b90a77d2885fb0429e8a21ab72fc73ca6f8b401	2016-01-05T05:09:10Z
 gopkg.in/check.v1	git	64131543e7896d5bcc6bd5a76287eb75ea96c673	2014-10-24T13:38:53Z
+gopkg.in/macaroon.v1	git	ab3940c6c16510a850e1c2dd628b919f0f3f1464	2015-01-21T11:42:31Z
```
Comment (mvo5) [inline dependencies.tsv]: Please also add a new build-dependency to debian/control.

Comment (matiasb) [inline dependencies.tsv]: Hmm... it looks like there is no package for macaroon.v1, what should I do in that case?

Comment (matiasb) [inline dependencies.tsv]: Related, https://bugs.launchpad.net/ubuntu/+source/juju-core/+bug/1508120/
(it seems in juju the macaroon package is embedded in the project).

**Comment by chipaca** [comment]
retest this please

File: dependencies.tsv (hunk 1 above)
Comment (mvo5) [inline dependencies.tsv]: We use a different policy (currently) from juju, so currently we need to create a macaroon.v1 package. We should discuss in the standup if we want to change that or not. I personally think we should keep the policy of packaging the dependencies for now.

File: overlord/auth/auth.go [hunk 2]
```diff
@@ -161,6 +165,62 @@ func newMacaroonAuthenticator(macaroon string, discharges []string) *MacaroonAut
...
+	}
+	err = m.UnmarshalBinary(decoded)
+	if err != nil {
+		return nil, err
+	}
+	return &m, nil
+}
+
+// UbuntuoneCaveatForDischarge returns the 3rd party caveat from the macaroon to be discharged by Ubuntuone
+func UbuntuoneCaveatForDischarge(m *macaroon.Macaroon) (string, error) {
```
Comment (niemeyer) [inline overlord/auth/auth.go]: Would it be fair to call this one `LoginCaveatID`? Is that what it is?

File: store/auth.go [hunk 3]
```diff
@@ -31,6 +31,8 @@ var (
 	// MyAppsPackageAccessAPI points to MyApps endpoint to get a package access macaroon
 	MyAppsPackageAccessAPI = myappsAPIBase + "api/2.0/acl/package_access/"
 	ubuntuoneAPIBase       = authURL()
+	// UbuntuoneLocation is the Ubuntuone location as defined in the store macaroon
+	UbuntuoneLocation = authLocation()
```
Comment (niemeyer) [inline store/auth.go]: Why so much indirection for this variable?  Isn't that a hardcoded caveat ID that we cannot touch?  If so, can we please have this in a const named `loginCaveatLocation` right next to that function?

Comment (niemeyer) [inline store/auth.go]: Well, maybe it does need to change.. please see below.

File: store/store.go [hunk 4]
```diff
@@ -124,11 +124,15 @@ func cpiURL() string {
 	return "https://search.apps.ubuntu.com/api/v1/"
 }
 
-func authURL() string {
+func authLocation() string {
 	if os.Getenv("SNAPPY_USE_STAGING_CPI") != "" {
-		return "https://login.staging.ubuntu.com/api/v2"
+		return "login.staging.ubuntu.com"
```
Comment (niemeyer) [inline store/store.go]: Okay, so it does change. Nevermind, sorry.

**Comment by niemeyer** [comment]
Please don't merge this until the packaging situation of macaroon.v1 is sorted out, so it doesn't block the SRU.

**Comment by chipaca** [comment]
:+1:

**Comment by pete-woods** [comment]
This looks good to me now 👍

**Pull request merged**